flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/security/token/HiveDelegationTokenProvider.java

```diff
@@ -0,0 +1,148 @@
+        /**
+         * The general rule how a provider/receiver must behave is the following: The provider and
+         * the receiver must be added to the classpath together with all the additionally required
+         * dependencies.
+         *
+         * <p>This null check is required because the Hive provider is always on classpath but Hive
+         * jars are optional. Such case configuration is not able to be loaded. This construct is
+         * intended to be removed when Hive provider/receiver pair can be externalized (namely if a
+         * provider/receiver throws an exception then workload must be stopped).
+         */
+        if (hiveConf == null) {
+            LOG.debug(
+                    "Hive is not available (not packaged with this application), hence no "
+                            + "tokens will be acquired.");
+            return false;
+        }
+        try {
+            if (!HadoopUtils.isKerberosSecurityEnabled(UserGroupInformation.getCurrentUser())) {
+                return false;
+            }
+        } catch (IOException e) {
+            LOG.debug("Hadoop Kerberos is not enabled.");
+            return false;
+        }
+        return !hiveConf.getTrimmed("hive.metastore.uris", "").isEmpty()
+                && kerberosLoginProvider.isLoginPossible(false);
+    }
+
+    @Override
+    public ObtainedDelegationTokens obtainDelegationTokens() throws Exception {
+        UserGroupInformation freshUGI = kerberosLoginProvider.doLoginAndReturnUGI();
+        return freshUGI.doAs(
+                (PrivilegedExceptionAction<ObtainedDelegationTokens>)
+                        () -> {
+                            Preconditions.checkNotNull(hiveConf);
+                            Hive hive = Hive.get((HiveConf) hiveConf);
+                            try {
+                                LOG.info("Obtaining Kerberos security token for Hive");
+
+                                String principal =
+                                        hiveConf.getTrimmed(
+                                                "hive.metastore.kerberos.principal", "");
+
+                                String tokenStr =
+                                        hive.getDelegationToken(
+                                                UserGroupInformation.getCurrentUser().getUserName(),
+                                                principal);
+                                Token<DelegationTokenIdentifier> hive2Token = new Token<>();
+                                hive2Token.decodeFromUrlString(tokenStr);
+
+                                Credentials credentials = new Credentials();
+                                credentials.addToken(TOKEN_ALIAS, hive2Token);
+
+                                return new ObtainedDelegationTokens(
+                                        HadoopDelegationTokenConverter.serialize(credentials),
+                                        Optional.empty());
```

Comment:
do we want to add renewal period here?

Reply:
Does hive token has validity period at all?

Reply:
There are several configurations around this:

- metastore.cluster.delegation.key.update-interval / hive.cluster.delegation.key.update-interval
- metastore.cluster.delegation.token.gc-interval / hive.cluster.delegation.token.gc-interval
- metastore.cluster.delegation.token.max-lifetime / hive.cluster.delegation.token.max-lifetime
- metastore.cluster.delegation.token.renew-interval / hive.cluster.delegation.token.renew-interval
- metastore.cluster.delegation.token.store.class / hive.cluster.delegation.token.store.class

See: https://github.com/apache/hive/blob/master/standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java#L743-L753
https://community.cloudera.com/t5/Support-Questions/How-to-Renew-Hive-Delegation-Token-Period/td-p/193509

Reply:
One additional note. This config values might not be in the hive-site.xml by default, as this is a metastore side configuration, and usually these are only set in the HMS config, but not in the HS2/client configs.
We might want to highlight it in some documentation/comments whatever

Reply:
Can we get renewInterval in the following way instead of reading from the configuration file, although this way is an approximate value @pvary 

```java
  long newExpiration = hive.getMSC().renewDelegationToken(tokenStr);
  String tokenKind = tokenIdentifier.getKind().toString();
  interval = newExpiration - getIssueDate(clock, tokenKind, tokenIdentifier);
```